File: fasolt.ios/Fasolt/ViewModels/StudyViewModel.swift
import Foundation

@MainActor
protocol StudyCardSource: AnyObject {
    func fetchDueCards(deckId: String?, limit: Int) async throws -> [DueCardDTO]
    func fetchCustomCards(deckId: String) async throws -> [DueCardDTO]
    func setSuspended(cardId: String, isSuspended: Bool) async throws
    func rateCard(cardId: String, rating: String) async throws -> RateCardResponse?
}

extension CardRepository: StudyCardSource {}

@MainActor
@Observable
final class StudyViewModel {
    enum SessionState {
        case idle, loading, studying, flipped, summary
    }

    var state: SessionState = .idle
    var errorMessage: String?
    var ratingError: String?
    var cards: [DueCardDTO] = []
    var currentIndex: Int = 0
    var isFlipped: Bool = false
    var ratingsCount: [String: Int] = ["again": 0, "hard": 0, "good": 0, "easy": 0]
    var cardsStudied: Int = 0
    var failedRatings: Int = 0
    var skippedCount: Int = 0
    var suspendedCount: Int = 0
    var mode: StudyMode = .normal
    private(set) var isRating = false

    var notificationService: NotificationService?

    private var hasRequestedPermission: Bool {
        get { UserDefaults.standard.bool(forKey: "hasRequestedNotificationPermission") }
        set { UserDefaults.standard.set(newValue, forKey: "hasRequestedNotificationPermission") }
    }

    private let cardRepository: any StudyCardSource

    init(cardRepository: any StudyCardSource) {
        self.cardRepository = cardRepository
    }

    var currentCard: DueCardDTO? {
        guard currentIndex < cards.count else { return nil }
        return cards[currentIndex]
    }

    var progress: Double {
        guard !cards.isEmpty else { return 0 }
        return Double(currentIndex) / Double(cards.count)
    }

    var totalCards: Int { cards.count }

    func startSession(deckId: String? = nil, mode: StudyMode = .normal) async {
        state = .loading
        errorMessage = nil
        self.mode = mode
        do {
            switch mode {
            case .normal:
                cards = try await cardRepository.fetchDueCards(deckId: deckId, limit: 50)
            case .cram:
                guard let deckId else {
                    errorMessage = "A deck is required for custom study."
                    state = .idle
                    return
                }
                cards = try await cardRepository.fetchCustomCards(deckId: deckId)
            }
            if cards.isEmpty {
                state = .summary
            } else {
                currentIndex = 0
                isFlipped = false
                cardsStudied = 0
                failedRatings = 0
                skippedCount = 0
                suspendedCount = 0
                ratingsCount = ["again": 0, "hard": 0, "good": 0, "easy": 0]
                state = .studying
            }
        } catch {
            errorMessage = "Could not load cards. Check your connection."
            state = .idle
        }
    }

    func advance() {
        cardsStudied += 1
        currentIndex += 1
        isFlipped = false
        if currentIndex >= cards.count {
            state = .summary
        } else {
            state = .studying
        }
    }

    func flipCard() {
        isFlipped = true
        state = .flipped
    }

    func skipCard() {
        skippedCount += 1
        currentIndex += 1
        isFlipped = false
        if currentIndex >= cards.count {
            state = .summary
            if cardsStudied > 0 {
                requestNotificationPermissionIfNeeded()
            }
        } else {
            state = .studying
        }
    }

    func suspendCard() async {
        guard let card = currentCard else { return }
        do {
            try await cardRepository.setSuspended(cardId: card.id, isSuspended: true)
        } catch {
            // best-effort — still skip the card
        }
        suspendedCount += 1
        currentIndex += 1
        isFlipped = false
        if currentIndex >= cards.count {
            state = .summary
            if cardsStudied > 0 {
                requestNotificationPermissionIfNeeded()
            }
        } else {
            state = .studying
        }
    }

    func rateCard(_ rating: String) async {
        guard let card = currentCard, !isRating else { return }
        isRating = true
        defer { isRating = false }
        ratingError = nil

        // Advance state immediately (optimistic update) so UI feels responsive
        ratingsCount[rating, default: 0] += 1
        cardsStudied += 1
        currentIndex += 1
        isFlipped = false
        if currentIndex >= cards.count {
            state = .summary
            requestNotificationPermissionIfNeeded()
        } else {
            state = .studying
        }

        // Submit rating to server in background — offline queue handles failures
        do {
            _ = try await cardRepository.rateCard(cardId: card.id, rating: rating)
        } catch {
            failedRatings += 1
            ratingError = "Rating may not have been saved."
        }
    }

    private func requestNotificationPermissionIfNeeded() {
        guard !hasRequestedPermission else { return }
        hasRequestedPermission = true
        Task {
            await notificationService?.requestPermissionAndRegister()
        }
    }

    func exitSession() {
        state = .idle
    }
}

File: fasolt.android/app/src/main/java/com/fasolt/android/ui/study/StudyViewModel.kt
package com.fasolt.android.ui.study

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.fasolt.android.FasoltApplication
import com.fasolt.android.data.api.models.DueCardDto
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

/** Top-level UI states for the Study flow — mirrors iOS StudyViewModel.SessionState. */
sealed interface StudyUiState {
    data object Idle : StudyUiState
    data object Loading : StudyUiState
    data class Error(val message: String) : StudyUiState
    data class Studying(
        val cards: List<DueCardDto>,
        val currentIndex: Int,
        val isFlipped: Boolean,
        val isRating: Boolean,
        val ratingError: String? = null,
    ) : StudyUiState {
        val currentCard: DueCardDto? get() = cards.getOrNull(currentIndex)
        val totalCards: Int get() = cards.size
        val progress: Float
            get() = if (cards.isEmpty()) 0f else currentIndex.toFloat() / cards.size.toFloat()
    }

    data class Summary(
        val cardsStudied: Int,
        val ratingsCount: Map<String, Int>,
        val failedRatings: Int,
    ) : StudyUiState
}

/** Valid rating keys, matching the server's case-insensitive accept and iOS lowercase usage. */
object StudyRatings {
    const val AGAIN = "again"
    const val HARD = "hard"
    const val GOOD = "good"
    const val EASY = "easy"
    val ALL = listOf(AGAIN, HARD, GOOD, EASY)
}

class StudyViewModel(application: Application) : AndroidViewModel(application) {
    private val app = application as FasoltApplication

    private val _uiState = MutableStateFlow<StudyUiState>(StudyUiState.Idle)
    val uiState: StateFlow<StudyUiState> = _uiState.asStateFlow()

    // Session-scoped tallies — kept here so rating submissions can update incrementally.
    private var cardsStudied = 0
    private var failedRatings = 0
    private val ratingsCount = mutableMapOf(
        StudyRatings.AGAIN to 0,
        StudyRatings.HARD to 0,
        StudyRatings.GOOD to 0,
        StudyRatings.EASY to 0,
    )

    fun startSession(deckId: String?) {
        _uiState.value = StudyUiState.Loading
        resetTallies()
        viewModelScope.launch {
            runCatching { app.reviewRepository.due(deckId, 50) }
                .onSuccess { cards ->
                    if (cards.isEmpty()) {
                        _uiState.value = StudyUiState.Summary(0, ratingsCount.toMap(), 0)
                    } else {
                        _uiState.value = StudyUiState.Studying(
                            cards = cards,
                            currentIndex = 0,
                            isFlipped = false,
                            isRating = false,
                        )
                    }
                }
                .onFailure {
                    _uiState.value = StudyUiState.Error(
                        it.message ?: "Could not load cards. Check your connection.",
                    )
                }
        }
    }

    fun flip() {
        val current = _uiState.value as? StudyUiState.Studying ?: return
        if (!current.isFlipped) {
            _uiState.value = current.copy(isFlipped = true)
        }
    }

    fun rate(rating: String) {
        val current = _uiState.value as? StudyUiState.Studying ?: return
        if (current.isRating || !current.isFlipped) return
        val card = current.currentCard ?: return

        // Optimistic update — advance immediately so the UI feels responsive.
        ratingsCount[rating] = (ratingsCount[rating] ?: 0) + 1
        cardsStudied += 1
        val nextIndex = current.currentIndex + 1
        if (nextIndex >= current.cards.size) {
            _uiState.value = StudyUiState.Summary(
                cardsStudied = cardsStudied,
                ratingsCount = ratingsCount.toMap(),
                failedRatings = failedRatings,
            )
        } else {
            _uiState.value = current.copy(
                currentIndex = nextIndex,
                isFlipped = false,
                isRating = true,
                ratingError = null,
            )
        }

        // Submit in background.
        viewModelScope.launch {
            runCatching { app.reviewRepository.rate(card.id, rating) }
                .onFailure {
                    failedRatings += 1
                    val s = _uiState.value
                    if (s is StudyUiState.Studying) {
                        _uiState.value = s.copy(
                            isRating = false,
                            ratingError = "Rating may not have been saved.",
                        )
                    } else if (s is StudyUiState.Summary) {
                        _uiState.value = s.copy(failedRatings = failedRatings)
                    }
                }
                .onSuccess {
                    val s = _uiState.value
                    if (s is StudyUiState.Studying) {
                        _uiState.value = s.copy(isRating = false)
                    }
                }
        }
    }

    /** Restart the session against the same deck (or all decks). */
    fun studyAgain(deckId: String?) = startSession(deckId)

    private fun resetTallies() {
        cardsStudied = 0
        failedRatings = 0
        StudyRatings.ALL.forEach { ratingsCount[it] = 0 }
    }
}
